### ISSUE TITLE
`PyTorchFileRecorder` memory usage

### ISSUE THREAD

**Issue by laggui** [tracel-ai/burn#1270]
**Describe the bug**
Trying to load resnet-18 [pre-trained weights](https://download.pytorch.org/models/resnet18-f37072fd.pth) from torchvision, the memory usage rapidly increases to over 32GB which causes the process to get killed on my Ubuntu machine.

**To Reproduce**
1. Check out the [models/resnet](https://github.com/tracel-ai/models/tree/resnet/resnet-burn) branch
1. Download the [pre-trained weights](https://download.pytorch.org/models/resnet18-f37072fd.pth)
1. Modify the `burn-core` function [`remap`](https://github.com/tracel-ai/burn/blob/main/burn-core/src/record/serde/data.rs#L180) to allow multiple mappings as described in this PR #1269 (you can check out the branch associated to the PR)
1. Modify the [`Cargo.toml`](https://github.com/tracel-ai/models/blob/resnet/resnet-burn/Cargo.toml) for the resnet example to use the locally modified version of Burn with this change
1. Run the example `cargo run --release --example inference`
1. Watch the memory usage increase until the process is killed

**Expected behavior**
The memory usage should not be as bad as it is currently for such a small model.

**Comment by Nikaidou-Shinku** [comment]
I made a minimal reproducible example, I hope it helps.
<https://github.com/Nikaidou-Shinku/burn-1270-mre>

**Comment by antimora** [comment]
> I made a minimal reproducible example, I hope it helps.
> 
> <https://github.com/Nikaidou-Shinku/burn-1270-mre>
> 
> 

Thank you. I will check it out.

**Comment by laggui** [comment]
Based on the information provided by @Nikaidou-Shinku on discord and the MRE I found that I had a pattern that did not cover all layers. My fix for the pattern chain in #1269 did not reuse the new name, so some names would still not match and thus I'd run into the OOM issue we described. I'll open another PR for that.

I'm still running into an issue though but it seems to be related to something else. `pickle::read_all` is not returning all the tensors actually saved in the `.pth` file. I'm only getting the batch norm layers, no convolutional layers weights seem to be parsed correctly. Will investigate further.

**Comment by antimora** [comment]
I will be investing the root cause but now that @Nikaidou-Shinku pointed to another problem (wrong keys), I suspect the NestedValue creation is blown up because there is a loop or something like this during unflattening. I will try fixing the two issues: memory doubling and recursion.

**Comment by antimora** [comment]
I have figured out the root cause of the issue. It was because a struct field in a model was not present in pt file. This coincided with incorrect renaming fields. Thanks to @Nikaidou-Shinku for catching it. 

https://github.com/tracel-ai/burn/pull/1286 PR fixes it. Please see the changes section on how I fixed it.

**Comment by laggui** [comment]
Fixed with PR #1286
